atlasdb-commons/src/main/java/com/palantir/common/compression/LZ4CompressingInputStream.java

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright 2016 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.common.compression;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.zip.Checksum;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.ByteStreams;
+
+import net.jpountz.lz4.LZ4BlockOutputStream;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.xxhash.XXHashFactory;
+
+/**
+ * {@link InputStream} that wraps a delegate InputStream, compressing its
+ * contents as they are read. Compression is managed via a compressing
+ * {@link OutputStream}. Reads pull data as necessary from the delegate
+ * InputStream, write the uncompressed data through the OutputStream, then
+ * serve the read from the resulting compressed buffer.
+ */
+public final class LZ4CompressingInputStream extends BufferedDelegateInputStream {
+
+    private static final LZ4Compressor COMPRESSOR = LZ4Factory.fastestInstance().fastCompressor();
+    private static final int DEFAULT_SEED = 0x9747b28c;
+    private static final int DEFAULT_BLOCK_SIZE = 1 << 16; // 64 KB
+
+    private final LZ4BlockOutputStream compressingStream;
+    private final int blockSize;
+
+    // Flag to prevent writes to the internal buffer while the
+    // compressing stream is reading from it.
+    private boolean readInProgress;
+    // Flag to indicate whether this stream has been exhausted.
+    private boolean finished;
+
+    public LZ4CompressingInputStream(InputStream delegate) throws IOException {
+        this(delegate, DEFAULT_BLOCK_SIZE);
+    }
+
+    public LZ4CompressingInputStream(InputStream delegate, int blockSize) throws IOException {
+        super(delegate);
```

Comment:
Good catch - this was only needed with the previous implementation, where the decompressing stream needed to read the LZ4 header from the delegate stream before being able to determine the buffer size appropriately. allocateBuffer shouldn't be necessary at all anymore.